imxrt1060-hal/src/gpio.rs

```diff
@@ -212,6 +219,84 @@ where
     }
 }
 
+/// GPIO input interrupt configurations.
+///
+/// These configurations do not take effect until
+/// GPIO input interrupts are enabled.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(u32)]
+pub enum InterruptConfiguration {
+    // TODO isn't it a bit repetitive to have "Sensitive" on all here?
+    LowLevelSensitive = 0,
+    HighLevelSensitive = 1,
+    RisingEdgeSensitive = 2,
+    FallingEdgeSensitive = 3,
+    EitherEdgeSensitive = 4,
+}
+
+impl<P> GPIO<P, Input>
+where
+    P: Pin,
+{
+    /// Enable (`true`) or disable (`false`) interrupts for this GPIO input.
+    pub fn set_interrupt_enable(&mut self, enable: bool) {
+        cortex_m::interrupt::free(|_| unsafe {
+            ral::modify_reg!(ral::gpio, self.register_block(), IMR, |imr| if enable {
+                imr | self.mask()
+            } else {
+                imr & !self.mask()
+            })
+        });
+    }
+
+    /// Indicates if interrupts are (`true`) or are not (`false`) enabled for this GPIO input.
+    pub fn is_interrupt_enabled(&self) -> bool {
+        unsafe { ral::read_reg!(ral::gpio, self.register_block(), IMR) & self.mask() != 0u32 }
+    }
+
+    /// Set the interrupt configuration for this GPIO input.
+    pub fn set_interrupt_configuration(&mut self, interrupt_configuration: InterruptConfiguration) {
+        if InterruptConfiguration::EitherEdgeSensitive == interrupt_configuration {
+            unsafe {
+                ral::modify_reg!(ral::gpio, self.register_block(), EDGE_SEL, |edge_sel| {
+                    edge_sel | self.mask()
+                });
+            }
+        } else {
+            unsafe {
+                ral::modify_reg!(ral::gpio, self.register_block(), EDGE_SEL, |edge_sel| {
+                    edge_sel & !self.mask()
+                });
+            }
+            // TODO is it ok to assume the numeric value of the enum is the icr?
+            let icr = interrupt_configuration as u32;
+            let icr_offset = self.icr_offset();
+            let icr_modify = |reg| reg & !(0b11 << icr_offset) | (icr << icr_offset);
+            if <P as Pin>::Offset::USIZE < 16 {
+                unsafe {
+                    ral::modify_reg!(ral::gpio, self.register_block(), ICR1, icr_modify);
+                }
+            } else {
+                unsafe {
+                    ral::modify_reg!(ral::gpio, self.register_block(), ICR2, icr_modify);
+                }
+            }
+        }
+    }
+
+    /// Indicates whether this GPIO input triggered an interrupt.
+    pub fn is_interrupt_status(&self) -> bool {
+        unsafe { ral::read_reg!(ral::gpio, self.register_block(), ISR) & self.mask() != 032 }
+    }
+
+    /// Clear the interrupt status flag.
+    pub fn clear_interrupt_status(&mut self) {
+        // TODO is it okay to unconditonally write to clear, or do we need to
```

Comment:
Should we check whether the bit really is set first? Is it ok to call this when the bit is not clear?

Reply:
The ISR register is "write-1-clear" (W1C). Once this write completes, the bit is cleared in the register.

I think we can skip the check. The user can check if it's set using `is_interrupt_status()`. If the bit is 0, writing a 1 to clear the bit has no effect.